File: threads/c/thread.c
#include <stdio.h>
#include <pthread.h>
#include <unistd.h>

void* task(void* arg) {
    char* name = (char*)arg;
    pid_t pid = getpid();
    pthread_t tid = pthread_self();

    for (int i = 0; i < 100; i++) {
        printf("Child: Task %s, iteration %d, PID %d, Thread ID %lu\n",
               name, i, pid, (unsigned long)tid);
        sleep(1);
    }
    return NULL;
}

int main() {
    pthread_t thread1, thread2;
    char *name1 = "A";
    char *name2 = "B";

    printf("Parent: PID %d\n", getpid());

    // run in parallel (or concurrently depending on machine resources)
    pthread_create(&thread1, NULL, task, name1);
    pthread_create(&thread2, NULL, task, name2);

    // join thread to main thread
    pthread_join(thread1, NULL);
    pthread_join(thread2, NULL);

    sleep(100);

    printf("All threads completed\n");
    return 0;
}

File: threads/python/thread.py
import threading
import os
import time

def task(name, delay):
    for i in range(100):
        print(f"Child: Task {name}, Iteration {i}, PID {os.getpid()} Thread ID {threading.get_ident()}")
        time.sleep(delay)

def main():
    print(f"Parent: {os.getpid()}")

    # concurrent threads
    thread1 = threading.Thread(target= task, args=("A", 1))
    thread2 = threading.Thread(target= task, args=("A", 2))

    thread1.start()
    thread2.start()

    # join threads to main thread
    thread1.join()
    thread2.join()

    time.sleep(100)

    print("Oh Wait it doesnt work like that")
main()
